### ISSUE TITLE
Problem in sending multiple lines consecutively to vterm buffer 

### ISSUE THREAD

**Issue by shg** [akermu/emacs-libvterm#391]
I am writing an inferior mode using emacs-vterm and have a problem in sending buffer content to the vterm buffer.

For testing, I wrote the following function and tested sending the text below to the vterm buffer. I see there are some missing/duplicating text in output.txt.

I have never seen any error when sending a single line. This happens only when I send multiple lines using loop. I am not sure whether the issue here is of vterm or my code. I'd like to know the correct way to sending multiple lines.

```elisp
(defun my-vterm-send-buffer ()
  (interactive)
  (save-excursion
    (goto-char (point-min))
    (while (not (eobp))
      (let ((current-line (thing-at-point 'line t)))
	(with-current-buffer (get-buffer "vterm")
	  (vterm-send-string current-line))
        (forward-line)))))
```

```text
echo 00 abcde fghij klmno pqrst uvexyz ABCDE FGHIJ KLMNO PQRST UVWXYZ >|output.txt
echo 01 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ >>output.txt
echo 02 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ >>output.txt
echo 03 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ >>output.txt
...
echo 99 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ >>output.txt
```

**Comment by Sbozzolo** [comment]
I happy to hear that you are building something on with vterm.

Is there a technical reason why you are not sending multiple lines as a single string? In [vterm-extra](https://github.com/Sbozzolo/vterm-extra/) I send the entire content of the buffer in one go and that seems to work.

If I have to guess, the problem may be that you are sending strings too quickly with the while loop.

**Comment by shg** [comment]
I chose sending line by line because I wanted to allow the user to see that the buffer content is being sent. But, I just tested what you suggest and I found that sending the whole buffer at a time actually seems to make the same result as sending line by line. I will rewrite my code so. Thank you.

By the way, I had tested adding a wait in the loop. In the case of the example above, adding (sleep-for 0.01) eliminates the issue. However, the problem still occurs even with 0.1 sec wait with some other scripts. Adding wait is probably not a good solution here.

**Comment by shg** [comment]
Thank you. I have tested with the following code and still see the same problem.

```elisp
(defun my-vterm-send-buffer-2 ()
  (interactive)
  (let ((buffer-content (buffer-string)))
    (with-current-buffer (get-buffer "vterm")
      (vterm-send-string buffer-content))))
```

**Comment by Sbozzolo** [comment]
What is the buffer content of your test?
What is the output?

**Comment by Sbozzolo** [comment]
`vterm-extra` implements a logic very similar to the one you posted. I tried to use it with a 100-lines buffer with
```
echo num abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ >>output.txt
```
The output I find is what I expect.

**Comment by shg** [comment]
Thank you again. I tested with the same buffer content I posted here first (100 lines as you wrote without the omission), and an example output.txt I get is like the one attached below.

I guess the issue might be environment dependent.

macOS 10.15.6
emacs 27.1 (homebrew's emacs-mac)
bash 5.0.18 (homebrew's bash)

```text
00 abcde fghij klmno pqrst uvexyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
01 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
02 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
03 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
04 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
05 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
06 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
07 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
08 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
09 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
10 abcde fghij klmno pqrst uvexyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
11 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
12 abcde fghij klmno pqecho 12 abcde fghij klmno pqecho 12 abcde fghij klmno pqecho 12 abcde fghij klmnopqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
14 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
15 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
16 abcde fghij klmno pqecho 16 abcde fghij klmno pqecho 16 abcde fghij klmno pqecho 16 abcde fghij klmnopqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
18 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
19 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
20 abcde fghij klmno pqecho 20 abcde fghij klmno pqecho 20 abcde fghij klmno pqecho 20 abcde fghij klmnopqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
22 abcde fghij klmno pqecho 22 abcde fghij klmno pqecho 22 abcde fghij klmno pqecho 22 abcde fghij klmnopqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
24 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
25 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
26 abcde fghij klmno pqecho 26 abcde fghij klmno pqecho 26 abcde fghij klmno pqecho 26 abcde fghij klmnopqecho 26 abcde fghij klmno pqecho 26 abcde fghij klmno pqecho 26 abcde fghij klmnopqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
29 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
30 abcde fghij klmno pqrst uvexyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
31 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
32 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
33 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
34 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
35 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
36 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
37 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
38 abcde fghij klmno pqecho 38 abcde fghij klmno pqecho 38 abcde fghij klmno pqecho 38 abcde fghij klmno pqechovwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
40 abcde fghij klmno pqrst uvexyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
41 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
42 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
43 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
44 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
45 abcde fghij klmno pqecho 45 abcde fghij klmno pqecho 45 abcde fghij klmno pqecho 45 abcde fghij klmn pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
47 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
48 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
49 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
50 abcde fghij klmno pqrst uvexyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
51 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
52 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
53 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
54 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
55 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
56 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
57 abcde fghij klmno pqecho 57 abcde fghij klmno pqecho 57 abcde fghij klmno pqecho 57 abcde fghij klmno pqecho 57 abcde E FGHIJ KLMNO PQRST UVWXYZ
59 abcde fghij klmno pqecho 59 abcde fghij klmno pqecho 59 abcde fghij klmno pqecho 59 abcde fghij klmnopqrst uvexyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
61 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
62 abcde fghij klmno pqecho 62 abcde fghij klmno pqecho 62 abcde fghij klmno pqecho 62 abcde fghij klmnopqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
64 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
65 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
66 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
67 abcde fghij klmno pqecho 67 abcde fghij klmno pqecho 67 abcde fghij klmno pqecho 67 abcde fghij klmnopqecho 67 abcde fghij klmno pqecho 67 abcde fghij klmno pqecho 67 abcde fghij klmnopqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
70 abcde fghij klmno pqrst uvexyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
71 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
72 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
73 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
74 abcde fghij klmno pqecho 74 abcde fghij klmno pqecho 74 abcde fghij klmno pqecho 74 abcde fghij klmno pqecho wxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
76 abcde fghij klmno pqecho 76 abcde fghij klmno pqecho 76 abcde fghij klmno pqecho 76 abcde fghij klmnopqecho 76 abcde fghij klmno pqecho 76 abcde fghij klmno pqecho 76 abcde fghij klmnopqecho 76 abcde fghij klmno pqecho 76 abcde fghij klmno pqecho 76 abcde fghij klmnopqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
80 abcde fghij klmno pqrst uvexyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
81 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
82 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
83 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
84 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
85 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
86 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
87 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
88 abcde fghij klmno pqecho 88 abcde fghij klmno pqecho 88 abcde fghij klmno pqecho 88 abcde fghij klmno pqecho 88 abcde f FGHIJ KLMNO PQRST UVWXYZ
90 abcde fghij klmno pqrst uvexyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
91 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
92 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
93 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
94 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
95 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
96 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
97 abcde fghij klmno pqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
98 abcde fghij klmno pqecho 98 abcde fghij klmno pqecho 98 abcde fghij klmno pqecho 98 abcde fghij klmnopqrst uvwxyz ABCDE FGHIJ KLMNO PQRST UVWXYZ
```

**Comment by shg** [comment]
I quickly did the same test with term-mode and shell-mode. And I see the very similar issue with term-mode but not with shell-mode.

```elisp
(defun my-term-send-buffer ()
  (interactive)
  (let ((proc (get-buffer-process "*terminal*")))
    (term-send-region proc (point-min) (point-max))))

(defun my-shell-send-buffer ()
  (interactive)
  (let ((proc (get-buffer-process "*shell*")))
    (comint-proc-query proc (buffer-string))))
```

**Comment by shg** [comment]
I found the similar issue exists in pasting large text into Terminal.app or iTerm.app. I also found that iTerm.app has "Paste Slowly" command that is probably for avoiding this sort of issue.

So maybe the possible way to avoid this problem in sending large text is to send the text line by line or chunk by chunk with a certain amount of interval. I quickly tested using sleep-for with different intervals and found it's not simple. The required interval seems to depend on the program and the content (script) to be sent and if I choose a value which is sufficient for most scripts, pasting large text will take very long time. Maybe I need to write some program-specific logic that sends text line by line while monitoring whether the program is ready to accept the next line.

**Comment by Sbozzolo** [comment]
You can if a new prompt was displayed: in that case, I expect vterm to be ready to accept new input.

**Comment by shg** [comment]
Sorry for the delayed update. This was solved using the bracketed paste mode. I found that what I originally wanted to do can be achieved also using the bracketed paste mode.

By the way, waiting a prompt did not work well. Because the process (Julia) does not show prompt until a structure (a block) is closed. Just for information although I am not sure whether there's a case in which checking the vterm is ready or not is necessary.
